### PR TITLE
Feature/create page factory

### PR DESCRIPTION
**Pull request by at-hanhnguyen2**
create PageFactory

### CODE REVIEW

File: src/test/java/vn/asiantech/base/DriverBase.java [hunk 1]
```diff
@@ -71,4 +77,34 @@ protected final void waitForPageRedirected(final WebDriver driver, final String
     protected final void waitVisibilityOfElement(final WebDriver driver, final By element) {
         new WebDriverWait(driver, DEFAULT_TIME_OUT).until(ExpectedConditions.visibilityOfElementLocated(element));
     }
+
+    private static void initElements(WebDriver driver, Object page) {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverBase.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/0b5f951929be6d6c059aa3808e36369cfcf9dfd5/src/test/java/vn/asiantech/base/DriverBase.java#L81"></span>Parameter driver should be final.</td>
    </tr>
  </tbody>
</table>

Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverBase.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/0b5f951929be6d6c059aa3808e36369cfcf9dfd5/src/test/java/vn/asiantech/base/DriverBase.java#L81"></span>Parameter page should be final.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverBase.java [hunk 2]
```diff
@@ -71,4 +77,34 @@ protected final void waitForPageRedirected(final WebDriver driver, final String
...
+        initElements(new DefaultElementLocatorFactory(driver), page);
+    }
+
+    private static void initElements(ElementLocatorFactory factory, Object page) {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverBase.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/0b5f951929be6d6c059aa3808e36369cfcf9dfd5/src/test/java/vn/asiantech/base/DriverBase.java#L85"></span>Parameter factory should be final.</td>
    </tr>
  </tbody>
</table>

Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverBase.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/0b5f951929be6d6c059aa3808e36369cfcf9dfd5/src/test/java/vn/asiantech/base/DriverBase.java#L85"></span>Parameter page should be final.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverBase.java [hunk 3]
```diff
@@ -71,4 +77,34 @@ protected final void waitForPageRedirected(final WebDriver driver, final String
...
+        initElements(new DefaultFieldDecorator(factory), page);
+    }
+
+    private static void initElements(FieldDecorator decorator, Object page) {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverBase.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/0b5f951929be6d6c059aa3808e36369cfcf9dfd5/src/test/java/vn/asiantech/base/DriverBase.java#L89"></span>Parameter decorator should be final.</td>
    </tr>
  </tbody>
</table>

Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverBase.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/0b5f951929be6d6c059aa3808e36369cfcf9dfd5/src/test/java/vn/asiantech/base/DriverBase.java#L89"></span>Parameter page should be final.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverBase.java [hunk 4]
```diff
@@ -71,4 +77,34 @@ protected final void waitForPageRedirected(final WebDriver driver, final String
...
+        for (Class proxyIn = page.getClass(); proxyIn != Object.class; proxyIn = proxyIn.getSuperclass()) {
+            proxyFields(decorator, page, proxyIn);
+        }
+
+    }
+
+    private static void proxyFields(FieldDecorator decorator, Object page, Class<?> proxyIn) {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverBase.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/0b5f951929be6d6c059aa3808e36369cfcf9dfd5/src/test/java/vn/asiantech/base/DriverBase.java#L96"></span>Parameter decorator should be final.</td>
    </tr>
  </tbody>
</table>

Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverBase.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/0b5f951929be6d6c059aa3808e36369cfcf9dfd5/src/test/java/vn/asiantech/base/DriverBase.java#L96"></span>Parameter page should be final.</td>
    </tr>
  </tbody>
</table>

Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/DriverBase.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/0b5f951929be6d6c059aa3808e36369cfcf9dfd5/src/test/java/vn/asiantech/base/DriverBase.java#L96"></span>Parameter proxyIn should be final.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/PageFactory.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+import org.openqa.selenium.WebDriver;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * @author tien.hoang
+ */
+class PageFactory<T> {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/PageFactory.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/0b5f951929be6d6c059aa3808e36369cfcf9dfd5/src/test/java/vn/asiantech/base/PageFactory.java#L11"></span>Type Javadoc comment is missing an @param <T> tag.</T></td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/PageFactory.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    private Class<T> clazz;
+    private WebDriver driver;
+    private static Map<String, Object> pages = new HashMap<>();
+
+    PageFactory(WebDriver driver, Class<T> clazz) {
```
Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/PageFactory.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/0b5f951929be6d6c059aa3808e36369cfcf9dfd5/src/test/java/vn/asiantech/base/PageFactory.java#L16"></span>Parameter driver should be final.</td>
    </tr>
  </tbody>
</table>

Comment (at-vinhhuynh) [inline src/test/java/vn/asiantech/base/PageFactory.java]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/akaj-team/selenium-test/blob/0b5f951929be6d6c059aa3808e36369cfcf9dfd5/src/test/java/vn/asiantech/base/PageFactory.java#L16"></span>Parameter clazz should be final.</td>
    </tr>
  </tbody>
</table>

File: src/test/java/vn/asiantech/base/DriverBase.java [hunk 7]
```diff
@@ -71,4 +77,34 @@ protected final void waitForPageRedirected(final WebDriver driver, final String
...
+
+    private static void initElements(final ElementLocatorFactory factory, final Object page) {
+        initElements(new DefaultFieldDecorator(factory), page);
+    }
+
+    private static void initElements(final FieldDecorator decorator, final Object page) {
+        for (Class proxyIn = page.getClass(); proxyIn != Object.class; proxyIn = proxyIn.getSuperclass()) {
+            proxyFields(decorator, page, proxyIn);
+        }
+
```
Comment (at-hangtran) [inline src/test/java/vn/asiantech/base/DriverBase.java]: Please remove this line

File: src/test/java/vn/asiantech/base/PageFactory.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
+package vn.asiantech.base;
+
+import org.openqa.selenium.WebDriver;
+
```
Comment (at-hangtran) [inline src/test/java/vn/asiantech/base/PageFactory.java]: Please remove this line

File: src/test/java/vn/asiantech/base/PageFactory.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+            Object object = newClazz.getConstructor(WebDriver.class).newInstance(driver);
+            pages.put(newClazz.getName(), object);
+            return (T) object;
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+}
```
Comment (at-hangtran) [inline src/test/java/vn/asiantech/base/PageFactory.java]: Please add a break line

File: src/test/java/vn/asiantech/base/PageFactory.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+            System.out.println(this.getClass().getSimpleName() + ": Create instance for " + newClazz.getSimpleName());
+            Object object = newClazz.getConstructor(WebDriver.class).newInstance(driver);
+            pages.put(newClazz.getName(), object);
+            return (T) object;
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
```
Comment (at-hangtran) [inline src/test/java/vn/asiantech/base/PageFactory.java]: Please remove this line

**Pull request merged**